### PR TITLE
task: #159 Replace Jest with Vite and Vitest

### LINKED ISSUE #158: Remove Linaria mocks from Jest global setup
### Context

TL;DR The global mocks we currently setup in `scripts/jest/jest-setup.js` for Linaria's `css` and `styled` utilities interfere with React Testing Library's DOM queries because they result in non-standard tag names being used (e.g. `<mock-styled.textarea>`).

This blocks our ability to write effective unit tests for components, effectively restricting us to the sole use of DOM snapshots.

The mocks for Linaria were required with previous versions of Linaria, but they are no longer necessary. See #152 for more details.

### Work to be done

I see two main options here:

1. Like #157, we remove the mocks and do a big bang update for all the snapshots; or,
2. We defer this work in favour of switching to Vitest (at which point, the Jest mocks for Linaria won't be relevant, and the snapshot will be replaced with Vitest-written versions). This change is proposed in #159.

At this stage, I'm in favour of (2), but that's based on the assumption that we'll actually agree to switch from Jest to Vitest.

### LINKED ISSUE #159: Replace Jest with Vite + Vitest
### Context

TL;DR Jest is hard to configure so that the Linaria CSS extraction on components is performed during test runs. Switching to a Vite + Vitest combination is, by comparison, trivial to achieve. This CSS extraction is important when rendering components via React Testing Library (RTL), because without it, the correct styles are not applied to the component under test, undermining many RTL DOM queries.

More details on this issue can be found in #152.

### Considerations

While the setup of Vite and Vitest is trivial, it does have a number of important considerations.

1. @adrian-reapit reported in #152 that they've seen poor performance from Vitest in GitHub Actions. While the Console FE team hasn't had issues in our CI, we're currently using Jenkins, not GitHub Actions, so it's not immediately clear if we'll see performance problems running Vitest in Elements'.
2. Vitest does not provide global functions by default like Jest does. Rather, Vitest prefers explicit imports for utilities like `test`, `describe` and `vi` (the equivalent of `jest`). That said, if we don't like the idea of having to add explicit imports for test functions, we can use Vitest's `globals` option to enable them as globals as per Jest's behaviour.
3. Since Vitest is ESM first, setting up test doubles for a module's default export looks a little different with Vitest's `vi.mock` than it does with Jest's `jest.mock`. This is not a problem, just a matter of education for contributors.
4. There's no easy way (that I'm aware of) to allow an incremental migration of test files from Jest to Vitest, so I believe this will need to be big bang cut-over. The Console FE team has successfully done this across our significant number of packages that back Console Cloud, so I don't believe it will be too difficult.
5. If we use Vite + Vitest for our unit tests, we'll also be able to use the same Vite config to build Elements, instead of our current direct use of Rollup. I would consider this to be out-of-scope for this work and, if we go ahead with this proposal, would prefer to see us replace our Rollup config separately.

### Proposal

The following work would be done as a single, large PR:

1. Add Vite, Vitest and Happy DOM as dev dependencies, and setup the Vite + Vitest config;
2. Rewrite all tests to use `vi` instead of `jest` and import `test`, `describe`, etc from Vitest (unless we would prefer to turn on globals?);
3. Update our `test` package scripts to run Vitest instead of Jest;
4. Remove all Jest-related config.

**note:** If we see problems with [Happy DOM](https://github.com/capricorn86/happy-dom) (e.g. we need APIs that it doesn't support), we can just use [JSDOM](https://github.com/jsdom/jsdom) instead. Happy DOM is [faster](https://github.com/capricorn86/happy-dom/wiki/#performance) than JSDOM for some things at the moment but the Console FE team has needed JSDOM for some of our packages.

### PR DESCRIPTION
**Pull request by kurtdoherty**
### Context

TL;DR Jest is hard to configure so that the Linaria CSS extraction on components is performed during test runs. Switching to a Vite + Vitest combination is, by comparison, trivial to achieve. This CSS extraction is important when rendering components via React Testing Library (RTL), because without it, the correct styles are not applied to the component under test, undermining many RTL DOM queries.

More details on this issue can be found in #152.

## This PR

- Fixes #159;
- Fixes #158 (as Jest has been replaced);
- Replaces the use of Rollup with Vite and Jest with Vitest.
  - Vite has a much simpler configuration than Rollup;
  - Vitest uses the Vite config in `vite.config.ts`, so it leverages the same plugins as our production build;
  - Storybook will also use the Vite config in `vite.config.ts` when it builds the stories for our components.
- While Vitest and Jest have similar APIs, they are not identical, so the following changes were also necessary:
  - `jest.fn` -> `vi.fn`, `jest.mock` -> `vi.mock`
  - Vitest is ESM only, which means module mocking works differently. This did not require many changes though.

Other changes include:

- Since Vite and Vitest are ESM first tools, `"type": "module"` has been added to the package's manifest;
- To provide a more modern definition of the package's exports for consumers using modern tools like Vite, an export map has been added that defines the package's exports (ESM bundle, CJS bundle and style.css);
- The TypeScript config was updated to use `module: "Preserve"` (which implies `moduleResolution: "Bundler"`). This is necessary because we're building Elements with a bundler. A new build-specific `tsconfig.build.json` has also been introduced. This allows Vite to focus on emitting production JS files while we use `tsc` to emit types in parallel. It also allows us to reduce the work `tsc` does when emitting types by disabling type checking (we do type checking during the `yarn check` script, so there's no need to do it again in the `yarn build` script;
- The `/tokens` folder was moved inside the `/src` folder;
- Refactor of how coverage badges are generated; and,
- The `@/` alias originally configure in both the TS config and the Rollup config has been replaced with an [import map](https://nodejs.org/docs/latest/api/packages.html#imports) in the package's manifest. This provides us with a single source of truth for any internal module mappings we wish to maintain.

**IMPORTANT:** Due to the nature of these changes, it was not possible to incrementally migrate from Jest to Vitest. As such, this PR takes "big bang" approach. I have called attention to the most relevant changes throughout the PR, so you can likely stay on this overview tab and scan through those comments instead of trying to review all 152 files individually. The vast majority of the changeset will be snapshot updates which can largely be ignored.

### CODE REVIEW

File: src/components/chip/__tests__/__snapshots__/chip.test.tsx.snap [hunk 1]
```diff
@@ -1,33 +1,35 @@
-// Jest Snapshot v1, https://goo.gl/fbAQLP
+// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html
 
-exports[`Chip component should match a snapshot 1`] = `
+exports[`Chip component > should match a snapshot 1`] = `
 {
   "asFragment": [Function],
   "baseElement": <body>
```
Comment (kurtdoherty) [inline src/components/chip/__tests__/__snapshots__/chip.test.tsx.snap:6]: **note:** I have not swept through all the existing snapshot tests to update them to use `asFragment` instead of capturing the whole `render` result. I will deal with that via #157 once this PR is merged in.

File: tsconfig.json [hunk 2]
```diff
@@ -2,37 +2,27 @@
...
     "noImplicitAny": false,
     "forceConsistentCasingInFileNames": true,
     "strict": true,
-    "typeRoots": ["./node_modules/@types"],
     "sourceMap": true,
     "inlineSources": true,
     "resolveJsonModule": true,
     "skipLibCheck": true,
-    "module": "ESNext",
+    "module": "Preserve",
```
Comment (kurtdoherty) [inline tsconfig.json:16]: **note:** Because we're using a bundler, we need to use `module: Preserve` to ensure TypeScript deals with module imports in the same way as modern bundlers do (e.g. no need for a file extension)

File: tsconfig.json [hunk 3]
```diff
@@ -2,37 +2,27 @@
...
-    "types": ["jest", "node", "react"],
-    "baseUrl": "./",
-    "paths": {
-      "@/*": ["src/*"]
-    },
-    "lib": ["es2015", "es2017", "dom", "es5", "es6", "esnext"],
-    "outDir": "dist",
+    "declarationMap": true,
+    "types": ["vite/client", "vitest/globals", "vite-plugin-svgr/client"],
+    "lib": ["ES2022", "dom"],
```
Comment (kurtdoherty) [inline tsconfig.json:21]: **note:** ES2022 is a stable option. The inclusion of `esnext` alongside `es5` etc was not necessary.

File: tsconfig.json [hunk 4]
```diff
@@ -2,37 +2,27 @@
...
-    "module": "ESNext",
+    "module": "Preserve",
     "target": "ES2022",
     "declaration": true,
-    "moduleResolution": "node",
-    "types": ["jest", "node", "react"],
-    "baseUrl": "./",
-    "paths": {
-      "@/*": ["src/*"]
-    },
```
Comment (kurtdoherty) [inline tsconfig.json:24]: **note:** This `paths` config is replaced with the import map in our package manifest.

File: tsconfig.json [hunk 5]
```diff
@@ -2,37 +2,27 @@
...
-    "lib": ["es2015", "es2017", "dom", "es5", "es6", "esnext"],
-    "outDir": "dist",
+    "declarationMap": true,
+    "types": ["vite/client", "vitest/globals", "vite-plugin-svgr/client"],
+    "lib": ["ES2022", "dom"],
+    "outDir": "dist/types",
+    "rootDir": "src",
```
Comment (kurtdoherty) [inline tsconfig.json:23]: **note:** `rootDir` is required to help TypeScript understand how to resolve the import maps in our package manifest

**Comment by kurtdoherty** [comment]
@willmcvay @adrian-reapit Apologies for the huge PR, but there's no avoiding all the snapshot updates and Vitest changes. As the PR description points out, I've highlighted all the interesting this to be aware of via the comments above, so there shouldn't be any need for you to view individual files unless there's specific things you need to check. Please don't both reviewing all 152 files individually; most of the changes are snapshot/test updates that are boilerplate.

**Review by willmcvay — APPROVED** [review]
It's a pretty big PR so hard to comment too deeply - probably needs merging and bedding in. LGTM 👍🏻
